system/nxcamera/nxcamera.c

```diff
@@ -0,0 +1,905 @@
+  if (ret < 0)
+    {
+      verr("VIDIOC_REQBUFS failed: %d\n", -ret);
+      return ret;
+    }
+
+  if (req.count < 2)
+    {
+      verr("VIDIOC_REQBUFS failed: not enough buffers\n");
+      return -ENOMEM;
+    }
+
+  pcam->nbuffers = req.count;
+  pcam->bufs = malloc(req.count * sizeof(*pcam->bufs));
+  pcam->buf_sizes = malloc(req.count * sizeof(*pcam->buf_sizes));
+  if (!pcam->bufs)
+    {
+      verr("Cannot allocate buffer pointers\n");
+      return -ENOMEM;
+    }
+
+  /* VIDIOC_QBUF enqueue buffer */
+
+  for (i = 0; i < req.count; i++)
+    {
+      memset(&buf, 0, sizeof(buf));
+      buf.type = req.type;
+      buf.memory = V4L2_MEMORY_MMAP;
+      buf.index = i;
+      ret = ioctl(pcam->capture_fd, VIDIOC_QUERYBUF, (uintptr_t)&buf);
+      if (ret < 0)
+        {
+          verr("VIDIOC_QUERYBUF failed: %d\n", -ret);
+          return ret;
+        }
+
+      pcam->bufs[i] = mmap(NULL, buf.length,
+                           PROT_READ | PROT_WRITE, MAP_SHARED,
+                           pcam->capture_fd, buf.m.offset);
+      if (pcam->bufs[i] == MAP_FAILED)
+        {
+          verr("MMAP failed\n");
+          return ret;
+        }
+
+      pcam->buf_sizes[i] = buf.length;
+    }
+
+  /* Create a message queue for the loopthread */
+
+  attr.mq_maxmsg  = req.count + 8;
+  attr.mq_msgsize = sizeof(struct video_msg_s);
+  attr.mq_curmsgs = 0;
+  attr.mq_flags   = 0;
+
+  snprintf(pcam->mqname, sizeof(pcam->mqname), "/tmp/%lx",
+           (unsigned long)((uintptr_t)pcam) & 0xffffffff);
+
+  pcam->mq = mq_open(pcam->mqname, O_RDWR | O_CREAT | O_NONBLOCK, 0644,
+                     &attr);
+  if (pcam->mq == (mqd_t) -1)
+    {
+      /* Unable to open message queue! */
+
+      ret = -errno;
+      auderr("ERROR: mq_open failed: %d\n", ret);
+      goto err_out;
+    }
+
+  /* Check if there was a previous thread and join it if there was
+   * to perform clean-up.
+   */
+
+  if (pcam->loop_id != 0)
+    {
+      pthread_join(pcam->loop_id, &value);
+    }
+
+  pthread_attr_init(&tattr);
+  sparam.sched_priority = sched_get_priority_max(SCHED_FIFO) - 9;
+  pthread_attr_setschedparam(&tattr, &sparam);
+  pthread_attr_setstacksize(&tattr, CONFIG_NXCAMERA_LOOPTHREAD_STACKSIZE);
+
+  /* Add a reference count to the looper for the thread and start the
+   * thread.  We increment for the thread to avoid thread start-up
+   * race conditions.
+   */
+
+  nxcamera_reference(pcam);
+  ret = pthread_create(&pcam->loop_id, &tattr, nxcamera_loopthread,
+                       (pthread_addr_t)pcam);
+  pthread_attr_destroy(&tattr);
+  if (ret != OK)
+    {
+      ret = -ret;
+      verr("ERROR: Failed to create loopthread: %d\n", ret);
+      goto err_out;
+    }
+
+  /* Name the thread */
+
+  pthread_setname_np(pcam->loop_id, "nxcameraloop");
+  return OK;
+
+err_out:
+  close(pcam->display_fd);
+  pcam->display_fd = -1;
+  close(pcam->capture_fd);
+  pcam->capture_fd = -1;
+  free(pcam->bufs);
```

Comment:
shouldn't we free `pcam->buf_sizes` as well?

Reply:
fixed